Source/BSN.Commons/Infrastructure/MessageBroker/Kafka/KafkaEventAggregator.cs

```diff
@@ -0,0 +1,192 @@
+            string messageString = JsonConvert.SerializeObject(@event);
+
+            var message = new Message<Null, string>
+            {
+                Value = messageString
+            };
+            
+            return _producer.ProduceAsync(eventName, message);
+        }
+
+        /// <inheritdoc />
+        public bool HasSubscriptionForEvent<TEvent, TEventDataModel>() where TEvent : IEvent<TEventDataModel> where TEventDataModel : IEventDataModel
+        {
+            return _subscriptionManager.HasSubscriptionsForEvent<TEvent>();
+        }
+
+        /// <inheritdoc />
+        public void Subscribe<TEvent, TEventDataModel>(IEventReceiver eventReceiver) where TEvent : IEvent<TEventDataModel> where TEventDataModel : IEventDataModel
+        {
+            try
+            {
+                _subscriptionManager.AddSubscription<TEvent>(eventReceiver);
+                
+                CreateConsumerForEvent<TEvent, TEventDataModel>(eventReceiver);
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e, $"Error subscribing to event {typeof(TEvent).FullName}.");
+                
+                _subscriptionManager.RemoveSubscription(typeof(TEvent).FullName);
+                
+                throw;
+            }
+        }
+
+        /// <inheritdoc />
+        public void UnSubscribe<TEvent, TEventDataModel>(IEventReceiver eventReceiver) where TEvent : IEvent<TEventDataModel> where TEventDataModel : IEventDataModel
+        {
+            string eventName = typeof(TEvent).Name;
+            
+            if (!_consumers.ContainsKey(eventName))
+            {
+                _logger.LogWarning($"Consumer for event {eventName} does not exist.");
+                return;
+            }
+            
+            _consumersCancellationTokenSources[eventName].Cancel();
+            
+            _consumers.Remove(eventName);
+            
+            _consumersCancellationTokenSources.Remove(eventName);
+            
+            _subscriptionManager.RemoveSubscription(eventName);
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            _producer?.Dispose();
+            
+            foreach (var consumer in _consumers)
+            {
+                _consumersCancellationTokenSources[consumer.Key].Cancel();
+                
+                consumer.Value.Dispose();
+            }
+        }
+        
+        private void CreateConsumerForEvent<TEvent, TEventDataModel>(IEventReceiver eventReceiver)
+            where TEvent : IEvent<TEventDataModel> where TEventDataModel : IEventDataModel
+        {
+            string eventName = typeof(TEvent).Name;
+
+            if (_consumers.ContainsKey(eventName))
+            {
+                _logger.LogWarning($"Consumer for event {eventName} already exists.");
+                return;
+            }
+
+            var consumer = new ConsumerBuilder<Null, string>(_consumerConfig).Build();
```

Comment:
I suggest use consumer factory instead of use directory

Reply:
done